### PR TITLE
First implementation of git appraise

### PR DESCRIPTION
**Pull request by sokolovstas**
Issue #189

This is first implementation of appraise review. I opened gitea code only this morning and to end of day I completed showing review list. 

Dear maintainers please see the code and tell me if I started implementation properly. I can continue to implement this tomorrow.

### CODE REVIEW

File: routers/repo/reviews.go
Comment (lunny) [inline routers/repo/reviews.go]: `2017 The Gitea`

File: routers/repo/reviews.go
Comment (lunny) [inline routers/repo/reviews.go]: Why not based on `Issue` like `Pull`, so that we can list all issues/pulls/reviews of one milestone.

File: routers/repo/reviews.go
Comment (lunny) [inline routers/repo/reviews.go]: IsPull needed?

**Review by lunny — CHANGES_REQUESTED** [review]
Almost it's great!

File: routers/repo/reviews.go
Comment (sokolovstas) [reply routers/repo/reviews.go]: I tried but I need that index field will be string, because appraise doesn't have unique in index

File: routers/repo/reviews.go
Comment (sokolovstas) [reply routers/repo/reviews.go]: Yes this may be removed, thank you

File: routers/repo/reviews.go
Comment (lunny) [reply routers/repo/reviews.go]: index's unique could be changed to index I think.

Comment (sokolovstas) [reply routers/repo/reviews.go]: Yes but it is int. In appraise that all we have is string
Revision: b9f1a3d0a1081a5f159ed90a3d413320890e0986

File: routers/repo/reviews.go
Comment (ethantkoenig) [inline routers/repo/reviews.go]: Local imports (e.g.  `code.gitea.io/gitea/foo/bar`) should come before non-stdlib external packages (e.g. `github.com/foo/bar`).

```
import (
	"strconv"
	"strings"
	"time"

	"code.gitea.io/gitea/models"
	"code.gitea.io/gitea/modules/base"
	"code.gitea.io/gitea/modules/context"

	"github.com/google/git-appraise/repository"
	"github.com/google/git-appraise/review"
)
```

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: routers/repo/reviews.go
Comment (sokolovstas) [reply routers/repo/reviews.go]: Added IsReview flag on Issue

File: routers/repo/reviews.go
Comment (makhov) [inline routers/repo/reviews.go:110]: Is it really needed to add the commented code to repo?

Comment (sokolovstas) [reply routers/repo/reviews.go:110]: I upload PR to other contributors and maintainers check my code for right implementation to exist code base. I will clear all commented lines on last commit.

File: routers/repo/reviews.go
Comment (lunny) [reply routers/repo/reviews.go]: So you can ignore index field and add revision in Reivew struct.

**Comment by lunny** [comment]
Maybe needs to insert or update one record to issue table when a review request pushed to git?

File: routers/repo/reviews.go
Comment (ojarjur) [inline routers/repo/reviews.go:408]: The logic for getting the commits in a review is a little more complicated than this because there are multiple states a review can be in (open, submitted, abandoned, etc).

I have a PR to git-appraise to make this a first-class part of the review API: https://github.com/google/git-appraise/pull/70

Comment (sokolovstas) [reply routers/repo/reviews.go:408]: @ojarjur thank you, I change usage of this function as soon as it was merged

**Comment by strk** [comment]
How can I test this ? I've a repository in which `git appraise list` shows there is a pending review for commit `09d5b03a69f3`, which I've pushed to gitea-hosted git repo with `git push gitea-remote --all`. Am I supposed to see the comments inline now, on the commit page ? Because I don't see it

**Comment by rikvdh** [comment]
I've just tested this out and it is a good start. Since we don't have line-annotations (#124) (like github does) perhaps just display line/file of a comment when it is present. Also rejection/acceptation is nice to have. Now it is just depicted as a regular comment.

Good work already, good to see the appraise comments in gitea already!

**Review by lunny — APPROVED** [review]

**Comment by strk** [comment]
@rikvdh as I could not see those comments instead, could you upload a screenshot ?

**Comment by ojarjur** [comment]
@strk, running 'git push --all' only pushes the branches, not the notes.

To push just the notes with review information, run 'git appraise push gitea-remote'

To mirror the entire repo, run 'git push --mirror'

**Comment by rikvdh** [comment]
Here are some screenshots including the latest commit.

request
The request uses the same template as an issue/pr, the review-request should probably get its own template and have `<name> requested review <time ago>` instead of `commented`, also all edit and delete buttons don't work and should be removed.

comments
All comments seem to get a empty code-snippet at the top, also issues and pull-request comments. When there is no line-file information the code-snippet box should be removed i guess.

When having a file/line reference perhaps add the name of the file, it is missing now and on big-reviews this can become very confusing.

The close and comment on the bottom seem to affect my pull-request 1 which is the only one in my test-repo.

Also abandoned reviews (git appraise abandon) should probably close the review, which are now unclose-able. The mutations like accept, reject and abandon are not clear.

Sorry for being so critical, I would really love this feature in Gitea, having a full reviewing system integrated is golden! 👍 Keep up the good work!

**Comment by bkcsoft** [comment]
How would this work if a read-only user wants to use `git-appraise` instead of ?
- Can they still push their review? Or can only users with write-access do that?
- If they can, how do you handle authorization? Adding that kind of complexity (checking for `notes/dev`-refs) sounds like inviting security issues 🙁 

Note: I'm very much against implementing something that only some people can use, if we go with `git-appraise` it should still work just like using the WebUI. In other words, if you can review in the WebUI you can also use `git-appraise`

My 2 cents 😉

**Comment by bkcsoft** [comment]
> `<name> requested review <time ago>` instead of `commented`

I would say `<name> reviewed <timeAgo>` 🙂

File: routers/repo/reviews.go
Comment (ojarjur) [reply routers/repo/reviews.go:408]: @sokolovstas That change has been merged into git-appraise now.

**Comment by ojarjur** [comment]
@bkcsoft 

> How would this work if a read-only user wants to use git-appraise instead of ?

I assume you mean how does this work if someone who does not have write access to a repo wants to use the git-appraise command line tool rather than the gitea UI.

The answer is that user would be able to pull reviews from the repository, but not push them to it. They could always push the review to their fork, but not to the upstream repository.

The git-appraise command-line tool uses the same auth mechanisms as git itself, so a user who cannot push directly to the repo also cannot directly push reviews.

One of the reasons for this design is that tools like gitea already define a way for outside contributors to authenticate and participate in a review, so git-appraise doesn't try to invent its own. Instead, whatever tool is responsible for auth (e.g. gitea) is also responsible for writing the review metadata to the repository.

So what I would expect is that the outside contributor creates their reviews and comments on them via the gitea UI, and the gitea server then writes them to the git-appraise formats.

> if we go with git-appraise it should still work just like using the WebUI. In other words, if you can review in the WebUI you can also use git-appraise

The important thing about git-appraise isn't the command-line tool but rather the storage formats.

Even if no one performs code reviews via the command line, having the review metadata stored directly in git is valuable.

**Comment by strk** [comment]
@ojarjur ok now I see the "Reviews" tab (I just didn't notice before).
Should these "Reviews" be associated to commits/pr rather than having their own tab ?

**Comment by roblabla** [comment]
@ojarjur would it not be possible to accept push to the git appraise reference by anyone that has read access to the repo ?

**Comment by lunny** [comment]
conflicted

**Comment by ojarjur** [comment]
> would it not be possible to accept push to the git appraise reference by anyone that has read access to the repo

@roblabla It should be possible to achieve what you want, but I would recommend a different implementation than that.

Reviews in git-appraise are just git objects the same way that commits are, and there is a well-established way for read-only contributors to push their own git objects. Specifically, they create a fork of the repository and then push to that fork. Those objects can then be mirrored into the upstream repository in a controlled way (e.g. like GitHub's 'refs/pull/...' refs).

The earlier question from @bkcsoft caused me to start thinking that we could build in a corresponding workflow for git-appraise, and I've created an issue for that [here](https://github.com/google/git-appraise/issues/71). Please add comments to that issue if the workflow described there doesn't match what you are looking for.

**Comment by bkcsoft** [comment]
> So what I would expect is that the outside contributor creates their reviews and comments on them via the gitea UI, and the gitea server then writes them to the git-appraise formats.

In that case I'm very much against this, since it adds unnecessary complexity to the system with no benefits what so ever to the users. The database is easy & fast to read/write, while git-appraise requires shelling out (slow) for everything and the added usability is hardly noticeable. 

IMO it would be better to store all this in the DB, and on pull we populate the refs required for git-appraise for those that want it. Obviously this would be read-only at first, until we've figured out how to read them on giteas side.

> Even if no one performs code reviews via the command line, having the review metadata stored directly in git is valuable.

@ojarjur On this we agree to some degree, which is why I mentioned populating `refs/notes/devtools/discuss` on `git appraise pull`. This means you will have your reviews stored offline 🙂

**Comment by bkcsoft** [comment]
> on pull we populate the refs required for git-appraise for those that want it.

Obviously this would be read-only at first, until we've figured out how to read them on giteas side.

**Comment by bkcsoft** [comment]
> Even if no one performs code reviews via the command line, having the review metadata stored directly in git is valuable.

@ojarjur On this we agree to some degree, which is why I mentioned populating `refs/notes/devtools/discuss` on `git pull`. This means that if you `git appraise pull` you will have your reviews stored offline 🙂

**Comment by lunny** [comment]
@sokolovstas conflicted.

**Comment by lunny** [comment]
@sokolovstas since #742 merged. All tab on repository page will be as a `Unit`. I think it asks this PR to add a new `UnitType` -> `UnitTypeReview`. Please rebase this one.

**Comment by lunny** [comment]
Just move this to v1.2

**Comment by olekukonko** [comment]
Any update on this functionality, please?

**Comment by sokolovstas** [comment]
Sorry but I don't have time to complete this, may be on next week

**Comment by lunny** [comment]
@sokolovstas if you could complete this PR before March 24th. I think it could be merged to v1.2 which will be planed release on April 24th.

**Comment by strk** [comment]
I'd love to see this completed too @sokolovstas -- any chance ?
Or anyone else would like to continue where he left ?

**Comment by lunny** [comment]
@sokolovstas if you have no time for this, maybe you can allow maintainers or others to continue the work so that you can check `Allow edits from maintainers.` on this page right bar bottom.

**Comment by sokolovstas** [comment]
@lunny This check box is already checked. For now I have very busy on main project and don't have free time. But will be very glad if someone can continue this work!

**Comment by strk** [comment]
@lunny are you going to give this a try ?

**Comment by lunny** [comment]
Maybe but I have to finish my PR first.

**Comment by Rushmead** [comment]
I've read through the comments and would love to help out with this and get it ready to be merged. Could someone make a list of exact things that need to be changed / added?

**Comment by strk** [comment]
Stuart I think you'll need to have an active role in this as there's
no great consensus about what the design should be. Personally I agree
about the UI using git repository itself for storage, rather than database
(as it makes federation much easier).

I guess your best first step would be using the proper UnitType and
rebasing / fixing conflicts.

**Comment by KangoV** [comment]
Any movement on this one? I don't know GO, but I could help with testing and feedback.

**Comment by lunny** [comment]
@KangoV it's not finished and no one continue with this now.

**Comment by lafriks** [comment]
I could probably work on this as I need it myself :)

**Comment by lunny** [comment]
@lafriks great news, please go ahead.

**Comment by strk** [comment]
@lafriks you are very welcome !

**Comment by strk** [comment]
This PR is in the wrong milestone, should be 1.3.0 (or 1.4.0, depending on @lafriks estimate)

**Comment by strk** [comment]
please disregard my previous comment, 1.x.x is clearly undefined but in the 1.x branch...

**Comment by lafriks** [comment]
@strk It's on my todo list just after more advanced branch protection and PR approval system :)

**Comment by jonasfranz** [comment]
@lafriks Is theire any further news about this PR or is it replaced by another PR-approval system in future?

**Comment by lafriks** [comment]
I started to work on this but I did not like it how it is currently implemented. I think it should be rewritten to be integrated with PRs

**Comment by lunny** [comment]
@lafriks I think rewrite is a better idea.

**Comment by techknowlogick** [comment]
Closing this due to PR approval/comment/review system in place via other PR.

**Pull request closed without merging**
